appium-tests/common/common.spec.js

```diff
@@ -0,0 +1,321 @@
+var contactsHelper = require('../helpers/contactsHelper');
+
+var MINUTE = 60 * 1000;
+var PLATFORM = global.PLATFORM;
+var UNORM = global.UNORM;
+
+describe('Contacts Android', function () {
+    var driver;
+    var webviewContext;
+    var callbackCount = 0;
+
+    function getNextCallbackId() {
+        return 'appium_callback_' + callbackCount++;
+    }
+
+    function saveScreenshotAndFail(error) {
+        fail(error);
+        return screenshotHelper
+            .saveScreenshot(driver)
+            .quit()
+            .then(function () {
+                return getDriver();
+            });
+    }
+
+    function getDriver() {
+        var getWebviewContext = function () {
+            return driver
+                .contexts()
+                .then(function (contexts) {
+                    var found = false;
+                    // take the last webview context
+                    for (var i = 0; i < contexts.length ; i++) {
+                        if (contexts[i].indexOf('WEBVIEW') >= 0) {
+                            webviewContext = contexts[i];
+                            found = true;
+                        }
+                    }
+                    if (!found) {
+                        // no webview context, the app is still loading
+                        return driver
+                            .sleep(10000)
+                            .then(getWebviewContext);
+                    }
+                });
+        };
+        driver = wdHelper.getDriver(PLATFORM);
+        return getWebviewContext();
+    }
+
+    function addContact(firstName, lastName) {
+        var contactName = contactsHelper.getContactName(firstName, lastName);
+        return driver
+            .context(webviewContext)
+            .setAsyncScriptTimeout(MINUTE)
+            .executeAsync(function(contactname, callback) {
+                navigator.contacts
+                    .create({ 'displayName': contactname.formatted, 'name': contactname, 'note': 'DeleteMe' })
+                    .save(callback, callback);
+            }, [contactName])
+            .then(function(result) {
+                if (result && result.hasOwnProperty('code')) {
+                    throw result;
+                }
+                return result;
+            });
+    }
+
+    function pickContact(name) {
+        var callbackId = getNextCallbackId();
+        return driver
+            .context(webviewContext)
+            .execute(function (cbId) {
+                var cbEl = document.createElement('div');
+                cbEl.id = cbId;
+                cbEl.style.display = 'none';
+                navigator.contacts.pickContact(function (contact) {
+                    cbEl.innerHTML = JSON.stringify(contact);
+                    document.body.appendChild(cbEl);
+                }, function (err) {
+                    cbEl.innerHTML = 'ERROR: ' + err;
+                    document.body.appendChild(cbEl);
+                });
```

Comment:
@riknoll and I thought about this approach and realized that this can be done in a cleaner way by using promises. Instead of creating a DOM element to store the result, you can wrap the `pickContact` call in a promise and save the promise on the `navigator` object. Then you can just call `.done()` on it using `execAsync` instead of using `waitForElementById`.